test/test_optim.py

```diff
@@ -35,6 +36,15 @@ def drosenbrock(tensor):
 class TestOptim(TestCase):
     exact_dtype = True
 
+    def setUp(self):
```

Comment:
What about modeling this like `exact_dtype` above, so it's easier for test writers to do

```
disable_tf32 = True/False
```

at the start of the class?

Reply:
Is the class level `exact_dtype` ever been used? I can only find it being used as an argument of `assertEqual`:

```python
gaoxiang@sunnyvale ~/pytorch-tf32 tf32-tests $ grep self.exact_dtype -r torch test/                                                                                                
test/test_sparse.py:        self.exact_dtype = True
gaoxiang@sunnyvale ~/pytorch-tf32 tf32-tests $ grep exact_dtype -r torch test/                                                                                                     
torch/testing/_internal/common_utils.py:    # If exact_dtype is true both tensors must have the same dtype.
torch/testing/_internal/common_utils.py:                        exact_dtype=True, exact_device=False) -> _compare_return_type:
torch/testing/_internal/common_utils.py:        # Checks dtype (if exact_dtype)
torch/testing/_internal/common_utils.py:        if exact_dtype and a.dtype is not b.dtype:
torch/testing/_internal/common_utils.py:        return self.assertEqual(*args, exact_dtype=False, **kwargs)
torch/testing/_internal/common_utils.py:                    equal_nan=True, exact_dtype=True, exact_device=False) -> None:
torch/testing/_internal/common_utils.py:                             exact_dtype=exact_dtype, exact_device=exact_device)
torch/testing/_internal/common_utils.py:                             exact_dtype=exact_dtype, exact_device=exact_device)
torch/testing/_internal/common_utils.py:                             exact_dtype=exact_dtype, exact_device=exact_device)
torch/testing/_internal/common_utils.py:                             exact_dtype=exact_dtype, exact_device=exact_device)
torch/testing/_internal/common_utils.py:                                                                 equal_nan=equal_nan, exact_dtype=exact_dtype,
torch/testing/_internal/common_utils.py:                                                                equal_nan=equal_nan, exact_dtype=exact_dtype,
torch/testing/_internal/common_utils.py:                                 msg=msg, exact_dtype=exact_dtype,
torch/testing/_internal/common_utils.py:                                     msg=msg, exact_dtype=exact_dtype,
torch/testing/_internal/common_utils.py:                                     exact_dtype=exact_dtype, exact_device=exact_device)
torch/testing/_internal/common_utils.py:                                     msg=msg, exact_dtype=exact_dtype,
torch/testing/_internal/common_utils.py:                                     exact_dtype=exact_dtype, exact_device=exact_device)
torch/testing/_internal/common_utils.py:                                     exact_dtype=exact_dtype, exact_device=exact_device)
torch/testing/_internal/common_utils.py:                                                         exact_dtype=exact_dtype,
torch/testing/_internal/common_utils.py:                                                         equal_nan=equal_nan, exact_dtype=exact_dtype,
torch/testing/_internal/common_utils.py:                                 msg=msg, exact_dtype=exact_dtype,
torch/testing/_internal/common_utils.py:                                 msg=msg, exact_dtype=exact_dtype,
torch/testing/_internal/common_utils.py:                                 exact_dtype=exact_dtype, exact_device=exact_device)
torch/testing/_internal/common_utils.py:                                 exact_dtype=exact_dtype, exact_device=exact_device)
test/test_bundled_inputs.py:                self.assertEqual(inp[0], samples[idx][0], exact_dtype=True)
test/test_spectral_ops.py:    exact_dtype = True
test/test_spectral_ops.py:                self.assertEqual(actual, expected, exact_dtype=(dtype is torch.complex128))
test/test_spectral_ops.py:                self.assertEqual(result, ref_result, atol=7e-6, rtol=0, msg='stft comparison against librosa', exact_dtype=False)
test/test_spectral_ops.py:            self.assertEqual(input, inverse, exact_dtype=True)
test/test_spectral_ops.py:                        atol=7e-6, rtol=0, exact_dtype=True)
test/test_spectral_ops.py:        self.assertEqual(i_original.repeat(1, 1), i_single, atol=1e-6, rtol=0, exact_dtype=True)
test/test_spectral_ops.py:        self.assertEqual(i_original.repeat(4, 1), i_multi, atol=1e-6, rtol=0, exact_dtype=True)
test/test_linalg.py:    exact_dtype = True
test/test_ops.py:    exact_dtype = True
test/test_ops.py:    exact_dtype = True
test/test_tensor_creation_ops.py:    exact_dtype = True
test/test_tensor_creation_ops.py:        self.assertEqual(res1, expected, exact_dtype=False)
test/test_tensor_creation_ops.py:        self.assertEqual(res1, expected, exact_dtype=False)
test/test_tensor_creation_ops.py:        self.assertEqual(res1, expected, exact_dtype=False)
test/test_tensor_creation_ops.py:            self.assertEqual(copy, source, exact_dtype=False)
test/test_tensor_creation_ops.py:        self.assertEqual(res1, expected, exact_dtype=False)
test/test_tensor_creation_ops.py:        self.assertEqual(res2, expected, exact_dtype=False)
test/test_tensor_creation_ops.py:        self.assertEqual(a, b, exact_dtype=False)
test/test_tensor_creation_ops.py:        self.assertEqual(a, b, exact_dtype=False)
test/test_tensor_creation_ops.py:        self.assertEqual(a, b, exact_dtype=False)
test/test_tensor_creation_ops.py:    exact_dtype = True
test/test_tensor_creation_ops.py:    exact_dtype = True
test/test_unary_ufuncs.py:    exact_dtype = True
test/test_unary_ufuncs.py:    def assertEqualHelper(self, actual, expected, msg, *, dtype, exact_dtype=True, **kwargs):
test/test_unary_ufuncs.py:            if exact_dtype:
test/test_optim.py:    exact_dtype = True
test/test_optim.py:    exact_dtype = True
test/test_sparse.py:        self.exact_dtype = True
test/test_sparse.py:            # exact_dtype if res.type is torch.float16.
test/test_sparse.py:                exact_dtype = False
test/test_sparse.py:                exact_dtype = True
test/test_sparse.py:            self.assertEqual(res, dense_x, exact_dtype=exact_dtype)
test/test_sparse.py:            self.assertEqual(res, safe_dense_x, exact_dtype=exact_dtype)
test/test_torch.py:                self.assertEqual(fn(x.to(dtype)), result, exact_dtype=False)
test/test_torch.py:                self.assertEqual(fn(x.to(dtype)), result, exact_dtype=False)
test/test_torch.py:                self.assertEqual(fn(x.to(other_dtype)), fn(x, dtype=other_dtype), exact_dtype=False)
test/test_torch.py:                    self.assertEqual(fn(x.to(mixed_dtype)), fn(x, dtype=mixed_dtype), exact_dtype=False)
test/test_torch.py:    exact_dtype = True
test/test_torch.py:                self.assertEqual(np_result, torch_result, exact_dtype=True)
test/test_torch.py:        result, debug_msg = self._compareTensors(a, b, exact_dtype=True)
test/test_torch.py:    def _comparetensors_helper(self, tests, device, dtype, equal_nan, exact_dtype=True, atol=1e-08, rtol=1e-05):
test/test_torch.py:                                                             exact_dtype=exact_dtype)
test/test_torch.py:                    self.assertEqual(res, ref, exact_dtype=False)
test/test_torch.py:        self.assertEqual(torch_result, numpy_result, exact_dtype=False)
test/test_torch.py:                exact_dtype=False)
test/test_torch.py:                self.assertEqual(indices, torch.take(input, indices), exact_dtype=False)
test/test_torch.py:                self.assertEqual(tensor_result, numpy_result, exact_dtype=False)
test/test_torch.py:                self.assertEqual(out, numpy_result, exact_dtype=False)
test/test_torch.py:                    self.assertEqual(torch.real(from_tensor), to_tensor, exact_dtype=False)
test/test_torch.py:                    self.assertEqual(from_tensor, torch.real(to_tensor), exact_dtype=False)
test/test_torch.py:                    self.assertEqual(torch.zeros_like(torch.imag(to_tensor)), torch.imag(to_tensor), exact_dtype=False)
test/test_torch.py:                    self.assertEqual(from_tensor, to_tensor, exact_dtype=False)
test/test_torch.py:                                            exact_dtype=True, with_keepdim=False):
test/test_torch.py:                                                    atol=atol, rtol=rtol, exact_dtype=exact_dtype)
test/test_torch.py:                                                    atol=atol, rtol=rtol, exact_dtype=exact_dtype)
test/test_torch.py:        exact_dtype = False if (IS_WINDOWS and is_integral(dtype)) else True
test/test_torch.py:                                                     atol=atol, rtol=rtol, exact_dtype=exact_dtype,
test/test_torch.py:    exact_dtype = True
test/test_torch.py:        self.assertEqual(torch.tensor(seq, dtype=dtype, device=device), reference, exact_dtype=False)
test/test_torch.py:    exact_dtype = True
test/test_torch.py:            self.assertEqual(x.real, new_real.real, exact_dtype=False)
test/test_torch.py:            self.assertEqual(x.imag, new_imag.real, exact_dtype=False)
test/test_torch.py:            self.assertEqual(x.real, new_real, exact_dtype=False)
test/test_torch.py:            self.assertEqual(x.imag, new_imag, exact_dtype=False)
test/test_torch.py:        self.assertEqual(cpu_tensor, device_tensor, atol=precision, rtol=0, exact_dtype=False)
test/test_torch.py:        self.assertEqual(cpu_args, device_args, atol=precision, rtol=0, exact_dtype=False)
test/test_torch.py:        self.assertEqual(cpu_result, device_result, atol=precision, rtol=0, exact_dtype=False)
test/test_torch.py:    exact_dtype = True
test/test_torch.py:    exact_dtype = True
test/test_torch.py:    exact_dtype = True
test/test_type_promotion.py:            self.assertEqual(res, expected_res, exact_dtype=True)
test/test_type_promotion.py:                self.assertEqual(out, expected_out, exact_dtype=True)
test/test_type_promotion.py:        self.assertEqual(result, t + t, exact_dtype=False)
test/test_type_promotion.py:        self.assertNotEqual(result, t.double() + t, exact_dtype=False)
test/test_type_promotion.py:        self.assertEqual(result, a / b, exact_dtype=False)
test/test_type_promotion.py:        self.assertNotEqual(result, a.double() / a, exact_dtype=False)
test/test_type_promotion.py:        self.assertEqual(result, a - b, exact_dtype=False)
test/test_type_promotion.py:        self.assertNotEqual(result, a.double() - b, exact_dtype=False)
test/test_nn.py:        self.assertEqual(out1, out2, atol=prec, rtol=0, exact_dtype=False)
test/test_nn.py:        self.assertEqual(input1.grad.data, input2.grad.data, atol=prec, rtol=0, exact_dtype=False)
test/test_nn.py:            self.assertEqual(out, ref_out, exact_dtype=False)
test/test_nn.py:            self.assertEqual(conv.weight.grad, ref_conv.weight.grad, exact_dtype=False)
test/test_nn.py:            self.assertEqual(conv.bias.grad, ref_conv.bias.grad, exact_dtype=False)
test/test_nn.py:            self.assertEqual(input.grad, ref_input.grad, exact_dtype=False)
test/test_jit.py:            self.assertEqual(torch.jit.script(foo)(1.), foo(1.), exact_dtype=True)
test/test_jit.py:            self.assertEqual(torch.jit.script(foo)(1.), foo(1.), exact_dtype=True)
test/test_jit.py:            self.assertEqual(torch.jit.script(foo)(1.), foo(1.), exact_dtype=True)
```

Reply:
@ngimel asks if this is really needed - how many tests in optim are actually affected by tf32?
@mruberry suggests that if we do want/need to support this pattern we put a bool on the class and update our TestCase class's setup and teardown methods to query that bool

Reply:
There is a `y = weight.mv(input)` in `_test_basic_cases_template`, so IIRC a lot tests are affected. But since we have disabled TF32 on mv now, let me rerun the test to see what happens

Reply:
Yes, this is no longer needed.